component/prometheus/kubernetes/podmonitors/crdManager.go

```diff
@@ -0,0 +1,189 @@
+	logger log.Logger
+	config *Arguments
+
+	discovery        *discovery.Manager
+	scraper          *scrape.Manager
+	cg               configGenerator
+	discoveryConfigs map[string]discovery.Configs
+	scrapeConfigs    map[string]*config.ScrapeConfig
+
+	mut sync.Mutex
+}
+
+func newManager(opts component.Options, logger log.Logger, cfg *Arguments) *crdManager {
+	return &crdManager{
+		opts:             opts,
+		logger:           logger,
+		config:           cfg,
+		discoveryConfigs: map[string]discovery.Configs{},
+		scrapeConfigs:    map[string]*config.ScrapeConfig{},
+	}
+}
+
+func (c *crdManager) apply() {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	err := c.discovery.ApplyConfig(c.discoveryConfigs)
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error applying discovery configs", "err", err)
+	}
+	scs := []*config.ScrapeConfig{}
+	for _, sc := range c.scrapeConfigs {
+		scs = append(scs, sc)
+	}
+	err = c.scraper.ApplyConfig(&config.Config{
+		ScrapeConfigs: scs,
+	})
+	if err != nil {
+		level.Error(c.logger).Log("msg", "error applying scrape configs", "err", err)
+	}
+	level.Debug(c.logger).Log("msg", "scrape config was updated")
+}
+
+func (c *crdManager) clearConfigs(kind string, ns string, name string) {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+	prefix := fmt.Sprintf("%s/%s/%s", kind, ns, name)
+	for k := range c.discoveryConfigs {
+		if strings.HasPrefix(k, prefix) {
+			delete(c.discoveryConfigs, k)
+			delete(c.scrapeConfigs, k)
+		}
+	}
+}
+
+func (c *crdManager) addPodMonitor(pm *v1.PodMonitor) {
+	for i, ep := range pm.Spec.PodMetricsEndpoints {
+		pmc, err := c.cg.generatePodMonitorConfig(pm, ep, i)
+		if err != nil {
+			level.Error(c.logger).Log("name", pm.Name, "err", err, "msg", "error generating scrapeconfig from podmonitor")
+			continue
+		}
+		c.mut.Lock()
+		c.discoveryConfigs[pmc.JobName] = pmc.ServiceDiscoveryConfigs
+		c.scrapeConfigs[pmc.JobName] = pmc
+		c.mut.Unlock()
+	}
+	c.apply()
+}
+func (c *crdManager) onAddPodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	level.Info(c.logger).Log("msg", "found pod monitor", "name", pm.Name)
+	c.addPodMonitor(pm)
+}
+func (c *crdManager) onUpdatePodMonitor(oldObj, newObj interface{}) {
+	pm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.addPodMonitor(newObj.(*v1.PodMonitor))
+}
+func (c *crdManager) onDeletePodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.apply()
+}
+
+func (c *crdManager) run(ctx context.Context) error {
+	config, err := c.config.restConfig()
+	if err != nil {
+		return errors.Wrap(err, "creating rest config")
+	}
+	promClientset, err := versioned.NewForConfig(config)
+	if err != nil {
+		return errors.Wrap(err, "creating prometheus clientset")
+	}
+
+	c.cg = configGenerator{
+		config: c.config,
+	}
+
+	for _, namespace := range c.config.Namespaces {
+		factory := informers.NewSharedInformerFactoryWithOptions(promClientset,
```

Comment:
Ideally we use the same library that loki.source.podlogs and loki.source.kubernetes_events use for informers (sigs.k8s.io/controller-runtime). If you don't want to switch to that library in this PR, can you add a TODO?